pybamm/expression_tree/operations/serialise.py

```diff
@@ -0,0 +1,384 @@
+                    node_dict["initial_condition"] = self.default(
+                        node.initial_condition
+                    )
+
+                return node_dict
+
+            if isinstance(node, pybamm.Event):
+                node_dict.update(node.to_json())
+                node_dict["expression"] = self.default(node._expression)
+                return node_dict
+
+            node_dict["json"] = json.JSONEncoder.default(self, node)  # pragma: no cover
+            return node_dict  # pragma: no cover
+
+    class _MeshEncoder(json.JSONEncoder):
+        """Converts PyBaMM meshes into a JSON-serialisable format"""
+
+        def default(self, node: dict):
+            node_dict = {"py/object": str(type(node))[8:-2], "py/id": id(node)}
+            if isinstance(node, pybamm.Mesh):
+                node_dict.update(node.to_json())
+
+                node_dict["sub_meshes"] = {}
+                for k, v in node.items():
+                    if len(k) == 1 and "ghost cell" not in k[0]:
+                        node_dict["sub_meshes"][k[0]] = self.default(v)
+
+                return node_dict
+
+            if isinstance(node, pybamm.SubMesh):
+                node_dict.update(node.to_json())
+                return node_dict
+
+            node_dict["json"] = json.JSONEncoder.default(self, node)  # pragma: no cover
+            return node_dict  # pragma: no cover
+
+    class _Empty:
+        """A dummy class to aid deserialisation"""
+
+        pass
+
+    class _EmptyDict(dict):
+        """A dummy dictionary class to aid deserialisation"""
+
+        pass
+
+    def save_model(
+        self,
+        model: pybamm.BaseModel,
+        mesh: pybamm.Mesh = None,
+        variables: pybamm.FuzzyDict = None,
+        filename: str = None,
+    ):
+        """Saves a discretised model to a JSON file.
+
+        As the model is discretised and ready to solve, only the right hand side,
+        algebraic and initial condition variables are saved.
+
+        Parameters
+        ----------
+        model : :class:`pybamm.BaseModel`
+            The discretised model to be saved
+        mesh : :class:`pybamm.Mesh` (optional)
+            The mesh the model has been discretised over. Not neccesary to solve
+            the model when read in, but required to use pybamm's plotting tools.
+        variables: :class:`pybamm.FuzzyDict` (optional)
+            The discretised model varaibles. Not necessary to solve a model, but
+            required to use pybamm's plotting tools.
+        filename: str (optional)
+            The desired name of the JSON file. If no name is provided, one will be
+            created based on the model name, and the current datetime.
+        """
+        if model.is_discretised is False:
+            raise NotImplementedError(
+                "PyBaMM can only serialise a discretised, ready-to-solve model."
+            )
+
+        model_json = {
+            "py/object": str(type(model))[8:-2],
```

Comment:
perhaps we should include the version of pybamm used to serialize the model? might need this when handling loading legacy models in later versions of pybamm

Reply:
Good idea, I'll add this.